avm/res/insights/private-link-scope/tests/e2e/waf-aligned/main.test.bicep

```diff
@@ -0,0 +1,196 @@
+targetScope = 'subscription'
+
+metadata name = 'WAF-aligned'
+metadata description = 'This instance deploys the module in alignment with the best-practices of the Azure Well-Architected Framework.'
+
+// ========== //
+// Parameters //
+// ========== //
+
+@description('Optional. The name of the resource group to deploy for testing purposes.')
+@maxLength(90)
+param resourceGroupName string = 'dep-${namePrefix}-insights.privatelinkscopes-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param resourceLocation string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'iplswaf'
+
+@description('Optional. A token to inject into the name of each resource. This value can be automatically injected by the CI.')
+param namePrefix string = '#_namePrefix_#'
+
+// ============ //
+// Dependencies //
+// ============ //
+
+// General resources
+// =================
+resource resourceGroup 'Microsoft.Resources/resourceGroups@2021-04-01' = {
+  name: resourceGroupName
+  location: resourceLocation
+}
+
+module nestedDependencies 'dependencies.bicep' = {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, resourceLocation)}-nestedDependencies'
+  params: {
+    virtualNetworkName: 'dep-${namePrefix}-vnet-${serviceShort}'
+    logAnalyticsWorkspaceName: 'dep-${namePrefix}-la-${serviceShort}'
+    location: resourceLocation
+  }
+}
+
+// ============== //
+// Test Execution //
+// ============== //
+var locationUpdated = toLower(replace(resourceLocation, ' ', ''))
+
+@batchSize(1)
+module testDeployment '../../../main.bicep' = [for iteration in [ 'init', 'idem' ]: {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, resourceLocation)}-test-${serviceShort}-${iteration}'
+  params: {
+    name: '${namePrefix}${serviceShort}001'
+    location: 'global'
+    accessModeSettings: {
+      exclusions: [
+        {
+          ingestionAccessMode: 'PrivateOnly'
+          queryAccessMode: 'PrivateOnly'
+          privateEndpointConnectionName: 'thisisatest'
+        }
+      ]
+      ingestionAccessMode: 'Open'
+      queryAccessMode: 'Open'
```

Comment:
In the object description is mentioned that PrivateOnly should be the value of both properties for security reasons and that they even default to that value if privateendpoint is set. Any specific reason for enforcing the other value here?

Reply:
So originally from the [MS Insights Private Link Scopes Ref Doco](https://learn.microsoft.com/en-us/azure/templates/microsoft.insights/privatelinkscopes?pivots=deployment-language-bicep#accessmodesettings), `ingestionAccessMode` and `queryAccessMode` are required properties. 

Now the intent (from our perspective) is to follow similar patterns we had around disabling public network access if PE is added into the resource. Hence the following was added into the resource property:

```bicep
    accessModeSettings: accessModeSettings ?? {
      ingestionAccessMode: empty(privateEndpoints) ? 'Open' : 'PrivateOnly'
      queryAccessMode: empty(privateEndpoints) ? 'Open' : 'PrivateOnly'
    }
```

On the UDT, I've set the following:
```bicep
  @description('Optional. Specifies the default access mode of ingestion through associated private endpoints in scope.')
  ingestionAccessMode: 'Open' | 'PrivateOnly'?

  @description('Optional. Specifies the default access mode of queries through associated private endpoints in scope.')
  queryAccessMode: 'Open' | 'PrivateOnly'?
```
So that we can make it `Optional` and default to `PrivateOnly` if PE is enabled from within the resource directly, for that it no longer becomes a `Required` parameter as per the Ref doc.

However, my deployments are failing now as its complaining that am I not passing these properties so it can default it to `PrivateOnly`. But the payload going to ARM is like this:

Test input:
```bicep
    accessModeSettings: {
      exclusions: [
        {
          ingestionAccessMode: 'PrivateOnly'
          queryAccessMode: 'PrivateOnly'
          privateEndpointConnectionName: 'thisisatest'
        }
      ]
      //ingestionAccessMode: 'Open'
      //queryAccessMode: 'Open'
    }
```

ARM response:
```pwsh
- properties.accessModeSettings.ingestionAccessMode is invalid, allowed values are: 'Open', 'PrivateOnly' (Code:BadRequest)
```

What gets passed as input to ARM:
```json
{
    "exclusions": [
        {
            "ingestionAccessMode": "PrivateOnly",
            "queryAccessMode": "PrivateOnly",
            "privateEndpointConnectionName": "thisisatest"
        }
    ]
}
```
Notice that `ingestionAccessMode` and `queryAccessMode` on the outer side are not even passed, although the intent is to get them in as `PrivateOnly` as per our resource logic.

Any advice on how this can be achieved? 
@eriqua @AlexanderSehr @ChrisSidebotham @segraef

Reply:
Ok I believe I landed in a good spot where the logic is working as intended, which is to set to `PrivateOnly` if Private Endpoints are passed. Also allow for overrides. 

```bicep
  properties: {
    accessModeSettings: !empty(privateEndpoints) ? {
      ingestionAccessMode: accessModeSettings.?ingestionAccessMode ?? 'PrivateOnly'
      queryAccessMode: accessModeSettings.?queryAccessMode ?? 'PrivateOnly'
      exclusions: accessModeSettings.?exclusions ?? []
    } : accessModeSettings ?? {
      ingestionAccessMode: accessModeSettings.?ingestionAccessMode ?? 'Open'
      queryAccessMode: accessModeSettings.?queryAccessMode ?? 'Open'
      exclusions: accessModeSettings.?exclusions ?? []
    }
  }
```

As for test cases:
- I've kept the `max` to expose the properties to show the overrides effect when being set to Open.
- In the `waf-aligned`, I've removed the access mode so that it will be set to `Private`.

Please let me know your thoughts so we can close this PR soon :)

Reply:
Fine with your logic.